feat: Non-blocking LSP requests for inlay hints and color hints

FalsePattern opened the pull request:
Proposed solution for https://github.com/redhat-developer/lsp4ij/issues/835

angelozerr commented on src/main/java/com/redhat/devtools/lsp4ij/internal/CompletableFutures.java:
```diff
@@ -109,6 +110,23 @@ public static void waitUntilDone(@Nullable CompletableFuture<?> future,
+     * @param pendingFutures the list to put the unfinished future into on timeout. If <code>null</code>,
+     *                       the timeout is retried indefinitely until successful.
+     */
+    public static boolean waitUntilDoneOrTimeout(@NotNull CompletableFuture<?> future,
```
Instead of having a new method, I think we should add an Integer timeout which can be null.

pendingFutures is a specific case that LSP inlay hints should manage.

angelozerr commented on src/main/java/com/redhat/devtools/lsp4ij/internal/CompletableFutures.java:
```diff
@@ -132,6 +150,10 @@ public static void waitUntilDone(@Nullable CompletableFuture<?> future,
                     throw new CancellationException("Some projects are indexing");
                 }
+                if (pendingFutures != null) {
+                    pendingFutures.add(future);
```
I wonder if it should be better to store pendingFutures in the editor data and use an HashSet to avoid adding the same future?

FalsePattern commented:
I've updated the code to use an overload for the waitUntilDone function instead, and replaced the `pendingFutures` system with sets. I'm not sure what you mean by "editor data".

angelozerr commented on src/main/java/com/redhat/devtools/lsp4ij/features/inlayhint/LSPInlayHintsProvider.java:
```diff
@@ -61,7 +61,7 @@ protected void doCollect(@NotNull PsiFile psiFile,
         try {
             // Wait until the future is finished and stop the wait if there are some ProcessCanceledException.
-            waitUntilDone(future, psiFile);
+            waitUntilDone(future, psiFile, 25);
```
Is 25 ms is enough? Perhaps we should set 500 ms to cover cases of ls which give  a fast LSP response?

angelozerr commented on src/main/java/com/redhat/devtools/lsp4ij/features/inlayhint/LSPInlayHintsProvider.java:
```diff
@@ -51,8 +52,7 @@ public class LSPInlayHintsProvider extends AbstractLSPDeclarativeInlayHintsProvi
     protected void doCollect(@NotNull PsiFile psiFile,
                              @NotNull Editor editor,
-                             @NotNull InlayTreeSink inlayHintsSink,
-                             @NotNull List<CompletableFuture<?>> pendingFutures) {
```
Please keep the `@NotNull List<CompletableFuture<?>> pendingFutures)`  parameters. Class which implement doCollect must don't know how pendingFutures is stored.

angelozerr commented:
@FalsePattern let me do some fix in an other PR.

angelozerr commented:
Thanks so much @FalsePattern for your great idea.

I close this PR in favor of https://github.com/redhat-developer/lsp4ij/pull/857
